Upgrading to 4.6 breaks type safety for OR

@bd82 I did some testing today on `4.6.0`, and `RULE<T>` now works as expected 🎆 
However, the upgrade broke `OR()` 😢 as now it returns `any` regardless of the type of the alternatives being passed. Should we also change `OR`, `OrMethodOpts`, and `IAnyOrAlt` to be of type `<T>` as well? this will enable both scenarios:

```ts
// automatically inferred: no conflict. OR returns an IToken
this.OR([
    { ALT: () => this.CONSUME(Token1) },
    { ALT: () => this.CONSUME(Token2) },

// cannot automatically infer: user will have to annotate the method manually.
this.OR<IToken | MyRuleType>([
    { ALT: () => this.CONSUME(Token1) },
    { ALT: () => this.SUBRULE(this.MyRule) },
```

@OmarTawfik 've looked into this and the problem is that when I add the generic types back to the `OR` method and its related interfaces then whenever there are multiple return types among the alternatives the end user would be forced to provide the generic type, otherwise an error would occur.

```typescript
    private value = this.RULE("value", () => {
        this.OR([
            { ALT: () => this.CONSUME(StringLiteral) },
            { ALT: () => this.CONSUME(NumberLiteral) },
            { ALT: () => this.SUBRULE(this.object) },
            { ALT: () => this.SUBRULE(this.array) },
            { ALT: () => this.CONSUME(True) },
            { ALT: () => this.CONSUME(False) },
            { ALT: () => this.CONSUME(Null) }
        ])
    })
```
```
typescript/typescript_json.ts:91:17 - error TS2345: Argument of type '({ ALT: () => IToken; } | { ALT: () => CstNode; })[]' is not assignable to parameter of typ
e 'OrMethodOpts<IToken> | IAnyOrAlt<IToken>[]'.
  Type '({ ALT: () => IToken; } | { ALT: () => CstNode; })[]' is not assignable to type 'IAnyOrAlt<IToken>[]'.
    Type '{ ALT: () => IToken; } | { ALT: () => CstNode; }' is not assignable to type 'IAnyOrAlt<IToken>'.
      Type '{ ALT: () => CstNode; }' is not assignable to type 'IAnyOrAlt<IToken>'.
        Property 'GATE' is missing in type '{ ALT: () => CstNode; }' but required in type 'IOrAltWithGate<IToken>'.

 91         this.OR([
                    ~
 92             { ALT: () => this.CONSUME(StringLiteral) },
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
 98             { ALT: () => this.CONSUME(Null) }
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 99         ])
    ~~~~~~~~~
```

It would be nice to have type safety for the `OR` method. but not at the cost
of enforcing a great percentage of its users to explicitly provide the generic type.
(Even if it was desired that would be a breaking change).

Do you have a suggestion how to model this to avoid the breaking change?
I could not find any way to define a return type that is inferred from a **dynamic** number
of generic arguments, is there such a thing in TypeScript?

--- a/packages/chevrotain/api.d.ts
+++ b/packages/chevrotain/api.d.ts
@@ -370,61 +370,71 @@ declare abstract class BaseParser {
      *
      * @returns The result of invoking the chosen alternative.
      */
-    /* protected */ OR(altsOrOpts: IOrAlt[] | OrMethodOpts): any
+    /* protected */ OR<T>(altsOrOpts: IOrAlt<T>[] | OrMethodOpts<T>): T
+    /* protected */ OR(altsOrOpts: IOrAlt<any>[] | OrMethodOpts<any>): any
 
     /**
      * @see OR
      * @hidden
      */
-    /* protected */ OR1(altsOrOpts: IOrAlt[] | OrMethodOpts): any
+    /* protected */ OR1<T>(altsOrOpts: IOrAlt<T>[] | OrMethodOpts<T>): T
+    /* protected */ OR1(altsOrOpts: IOrAlt<any>[] | OrMethodOpts<any>): any
 
     /**
      * @see OR
      * @hidden
      */
-    /* protected */ OR2(altsOrOpts: IOrAlt[] | OrMethodOpts): any
+    /* protected */ OR2<T>(altsOrOpts: IOrAlt<T>[] | OrMethodOpts<T>): T
+    /* protected */ OR2(altsOrOpts: IOrAlt<any>[] | OrMethodOpts<any>): any
 
     /**
      * @see OR
      * @hidden
      */
-    /* protected */ OR3(altsOrOpts: IOrAlt[] | OrMethodOpts): any
+    /* protected */ OR3<T>(altsOrOpts: IOrAlt<T>[] | OrMethodOpts<T>): T
+    /* protected */ OR3(altsOrOpts: IOrAlt<any>[] | OrMethodOpts<any>): any
 
     /**
      * @see OR
      * @hidden
      */
-    /* protected */ OR4(altsOrOpts: IOrAlt[] | OrMethodOpts): any
+    /* protected */ OR4<T>(altsOrOpts: IOrAlt<T>[] | OrMethodOpts<T>): T
+    /* protected */ OR4(altsOrOpts: IOrAlt<any>[] | OrMethodOpts<any>): any
 
     /**
      * @see OR
      * @hidden
      */
-    /* protected */ OR5(altsOrOpts: IOrAlt[] | OrMethodOpts): any
+    /* protected */ OR5<T>(altsOrOpts: IOrAlt<T>[] | OrMethodOpts<T>): T
+    /* protected */ OR5(altsOrOpts: IOrAlt<any>[] | OrMethodOpts<any>): any
 
     /**
      * @see OR
      * @hidden
      */
-    /* protected */ OR6(altsOrOpts: IOrAlt[] | OrMethodOpts): any
+    /* protected */ OR6<T>(altsOrOpts: IOrAlt<T>[] | OrMethodOpts<T>): T
+    /* protected */ OR6(altsOrOpts: IOrAlt<any>[] | OrMethodOpts<any>): any
 
     /**
      * @see OR
      * @hidden
      */
-    /* protected */ OR7(altsOrOpts: IOrAlt[] | OrMethodOpts): any
+    /* protected */ OR7<T>(altsOrOpts: IOrAlt<T>[] | OrMethodOpts<T>): T
+    /* protected */ OR7(altsOrOpts: IOrAlt<any>[] | OrMethodOpts<any>): any
 
     /**
      * @see OR
      * @hidden
      */
-    /* protected */ OR8(altsOrOpts: IOrAlt[] | OrMethodOpts): any
+    /* protected */ OR8<T>(altsOrOpts: IOrAlt<T>[] | OrMethodOpts<T>): T
+    /* protected */ OR8(altsOrOpts: IOrAlt<any>[] | OrMethodOpts<any>): any
 
     /**
      * @see OR
      * @hidden
      */
-    /* protected */ OR9(altsOrOpts: IOrAlt[] | OrMethodOpts): any
+    /* protected */ OR9<T>(altsOrOpts: IOrAlt<T>[] | OrMethodOpts<T>): T
+    /* protected */ OR9(altsOrOpts: IOrAlt<any>[] | OrMethodOpts<any>): any
 
     /**
      * Parsing DSL method, that indicates a repetition of zero or more.
@@ -866,7 +876,7 @@ export declare class Parser extends BaseParser {
         name: string,
         implementation: (...implArgs: any[]) => T,
         config?: IRuleConfig<T>
-    ): (idxInCallingRule?: number, ...args: any[]) => T | any
+    ): (idxInCallingRule?: number, ...args: any[]) => T
 
     /**
      * Same as {@link Parser.RULE}, but should only be used in
@@ -877,7 +887,7 @@ export declare class Parser extends BaseParser {
         name: string,
         impl: (...implArgs: any[]) => T,
         config?: IRuleConfig<T>
-    ): (idxInCallingRule?: number, ...args: any[]) => T | any
+    ): (idxInCallingRule?: number, ...args: any[]) => T
 
     /**
      * The Parsing DSL Method is used by one rule to call another.
@@ -1823,13 +1833,13 @@ export interface DSLMethodOptsWithErr<T> extends DSLMethodOpts<T> {
     ERR_MSG?: string
 }
 
-export interface OrMethodOpts {
+export interface OrMethodOpts<T> {
     NAME?: string
     /**
      * The set of alternatives,
      * See detailed description in {@link Parser.OR}
      */
-    DEF: IOrAlt[]
+    DEF: IOrAlt<T>[]
     /**
      * A description for the alternatives used in error messages
      * If none is provided, the error message will include the names of the expected
@@ -1922,10 +1932,10 @@ export declare type GrammarAction<OUT> = () => OUT
  */
 export declare type IAnyOrAlt = any
 
-export interface IOrAlt {
+export interface IOrAlt<T> {
     NAME?: string
     GATE?: () => boolean
-    ALT: () => any
+    ALT: () => T
     /**
      * A Flag indicating that any ambiguities involving this
      * specific alternative Should be ignored.
@@ -1936,7 +1946,7 @@ export interface IOrAlt {
     IGNORE_AMBIGUITIES?: boolean
 }
 
-export interface IOrAltWithGate extends IOrAlt {
+export interface IOrAltWithGate<T> extends IOrAlt<T> {
     // TODO: deprecate this interface
 }
 

--- a/packages/chevrotain/src/parse/cst/cst_visitor.ts
+++ b/packages/chevrotain/src/parse/cst/cst_visitor.ts
@@ -183,9 +183,7 @@ export function validateRedundantMethods(
                         any
                     >visitorInstance.constructor)} CST Visitor\n` +
                     `There is no Grammar Rule corresponding to this method's name.\n` +
-                    `For utility methods on visitor classes use methods names that do not match /${
-                        validTermsPattern.source
-                    }/.`,
+                    `For utility methods on visitor classes use methods names that do not match /${validTermsPattern.source}/.`,
                 type: CstVisitorDefinitionError.REDUNDANT_METHOD,
                 methodName: prop
             })

--- a/packages/chevrotain/src/parse/parser/traits/gast_recorder.ts
+++ b/packages/chevrotain/src/parse/parser/traits/gast_recorder.ts
@@ -263,7 +263,7 @@ export class GastRecorder {
 
     orInternalRecord<T>(
         this: MixedInParser,
-        altsOrOpts: IAnyOrAlt[] | OrMethodOpts,
+        altsOrOpts: IAnyOrAlt[] | OrMethodOpts<unknown>,
         occurrence: number
     ): T {
         return recordOrProd.call(this, altsOrOpts, occurrence)

--- a/packages/chevrotain/src/parse/parser/traits/recognizer_api.ts
+++ b/packages/chevrotain/src/parse/parser/traits/recognizer_api.ts
@@ -264,43 +264,73 @@ export class RecognizerApi {
         return this.optionInternal(actionORMethodDef, 9)
     }
 
-    OR<T>(this: MixedInParser, altsOrOpts: IAnyOrAlt[] | OrMethodOpts): T {
+    OR<T>(
+        this: MixedInParser,
+        altsOrOpts: IAnyOrAlt[] | OrMethodOpts<unknown>
+    ): T {
         return this.orInternal(altsOrOpts, 0)
     }
 
-    OR1<T>(this: MixedInParser, altsOrOpts: IAnyOrAlt[] | OrMethodOpts): T {
+    OR1<T>(
+        this: MixedInParser,
+        altsOrOpts: IAnyOrAlt[] | OrMethodOpts<unknown>
+    ): T {
         return this.orInternal(altsOrOpts, 1)
     }
 
-    OR2<T>(this: MixedInParser, altsOrOpts: IAnyOrAlt[] | OrMethodOpts): T {
+    OR2<T>(
+        this: MixedInParser,
+        altsOrOpts: IAnyOrAlt[] | OrMethodOpts<unknown>
+    ): T {
         return this.orInternal(altsOrOpts, 2)
     }
 
-    OR3<T>(this: MixedInParser, altsOrOpts: IAnyOrAlt[] | OrMethodOpts): T {
+    OR3<T>(
+        this: MixedInParser,
+        altsOrOpts: IAnyOrAlt[] | OrMethodOpts<unknown>
+    ): T {
         return this.orInternal(altsOrOpts, 3)
     }
 
-    OR4<T>(this: MixedInParser, altsOrOpts: IAnyOrAlt[] | OrMethodOpts): T {
+    OR4<T>(
+        this: MixedInParser,
+        altsOrOpts: IAnyOrAlt[] | OrMethodOpts<unknown>
+    ): T {
         return this.orInternal(altsOrOpts, 4)
     }
 
-    OR5<T>(this: MixedInParser, altsOrOpts: IAnyOrAlt[] | OrMethodOpts): T {
+    OR5<T>(
+        this: MixedInParser,
+        altsOrOpts: IAnyOrAlt[] | OrMethodOpts<unknown>
+    ): T {
         return this.orInternal(altsOrOpts, 5)
     }
 
-    OR6<T>(this: MixedInParser, altsOrOpts: IAnyOrAlt[] | OrMethodOpts): T {
+    OR6<T>(
+        this: MixedInParser,
+        altsOrOpts: IAnyOrAlt[] | OrMethodOpts<unknown>
+    ): T {
         return this.orInternal(altsOrOpts, 6)
     }
 
-    OR7<T>(this: MixedInParser, altsOrOpts: IAnyOrAlt[] | OrMethodOpts): T {
+    OR7<T>(
+        this: MixedInParser,
+        altsOrOpts: IAnyOrAlt[] | OrMethodOpts<unknown>
+    ): T {
         return this.orInternal(altsOrOpts, 7)
     }
 
-    OR8<T>(this: MixedInParser, altsOrOpts: IAnyOrAlt[] | OrMethodOpts): T {
+    OR8<T>(
+        this: MixedInParser,
+        altsOrOpts: IAnyOrAlt[] | OrMethodOpts<unknown>
+    ): T {
         return this.orInternal(altsOrOpts, 8)
     }
 
-    OR9<T>(this: MixedInParser, altsOrOpts: IAnyOrAlt[] | OrMethodOpts): T {
+    OR9<T>(
+        this: MixedInParser,
+        altsOrOpts: IAnyOrAlt[] | OrMethodOpts<unknown>
+    ): T {
         return this.orInternal(altsOrOpts, 9)
     }
 

--- a/packages/chevrotain/src/parse/parser/traits/recognizer_engine.ts
+++ b/packages/chevrotain/src/parse/parser/traits/recognizer_engine.ts
@@ -753,12 +753,12 @@ export class RecognizerEngine {
 
     orInternalNoCst(
         this: MixedInParser,
-        altsOrOpts: IAnyOrAlt[] | OrMethodOpts,
+        altsOrOpts: IAnyOrAlt[] | OrMethodOpts<unknown>,
         occurrence: number
     ): any {
         let alts = isArray(altsOrOpts)
             ? (altsOrOpts as IAnyOrAlt[])
-            : (altsOrOpts as OrMethodOpts).DEF
+            : (altsOrOpts as OrMethodOpts<unknown>).DEF
 
         const laKey = this.getKeyForAutomaticLookahead(OR_IDX, occurrence)
         const laFunc = this.getLaFuncFromCache(laKey)
@@ -769,25 +769,25 @@ export class RecognizerEngine {
         }
         this.raiseNoAltException(
             occurrence,
-            (altsOrOpts as OrMethodOpts).ERR_MSG
+            (altsOrOpts as OrMethodOpts<unknown>).ERR_MSG
         )
     }
 
     orInternal<T>(
         this: MixedInParser,
-        altsOrOpts: IAnyOrAlt[] | OrMethodOpts,
+        altsOrOpts: IAnyOrAlt[] | OrMethodOpts<unknown>,
         occurrence: number
     ): T {
         let laKey = this.getKeyForAutomaticLookahead(OR_IDX, occurrence)
         let nestedName = this.nestedRuleBeforeClause(
-            <OrMethodOpts>altsOrOpts,
+            <OrMethodOpts<unknown>>altsOrOpts,
             laKey
         )
 
         try {
             let alts = isArray(altsOrOpts)
                 ? (altsOrOpts as IAnyOrAlt[])
-                : (altsOrOpts as OrMethodOpts).DEF
+                : (altsOrOpts as OrMethodOpts<unknown>).DEF
 
             const laFunc = this.getLaFuncFromCache(laKey)
             let altIdxToTake = laFunc.call(this, alts)
@@ -812,7 +812,7 @@ export class RecognizerEngine {
             }
             this.raiseNoAltException(
                 occurrence,
-                (altsOrOpts as OrMethodOpts).ERR_MSG
+                (altsOrOpts as OrMethodOpts<unknown>).ERR_MSG
             )
         } finally {
             if (nestedName !== undefined) {

--- a/packages/chevrotain/src/scan/lexer_errors_public.ts
+++ b/packages/chevrotain/src/scan/lexer_errors_public.ts
@@ -2,9 +2,7 @@ import { ILexerErrorMessageProvider, IToken } from "../../api"
 
 export const defaultLexerErrorProvider: ILexerErrorMessageProvider = {
     buildUnableToPopLexerModeMessage(token: IToken): string {
-        return `Unable to pop Lexer Mode after encountering Token ->${
-            token.image
-        }<- The Mode Stack is empty`
+        return `Unable to pop Lexer Mode after encountering Token ->${token.image}<- The Mode Stack is empty`
     },
 
     buildUnexpectedCharactersMessage(

--- a/packages/chevrotain/test/parse/cst_spec.ts
+++ b/packages/chevrotain/test/parse/cst_spec.ts
@@ -98,7 +98,7 @@ function defineTestSuit(recoveryMode) {
         })
 
         it("Can output a CST with labels in recovery", () => {
-            class CstTerminalParserWithLabels extends Parser {
+            class CstTerminalParserWithLabels extends CstParser {
                 constructor(input: IToken[] = []) {
                     super(ALL_TOKENS, {
                         outputCst: true,
@@ -128,12 +128,14 @@ function defineTestSuit(recoveryMode) {
             expect(tokenStructuredMatcher(cst.children.myLabel[0], A)).to.be
                 .true
             expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be.true
-            expect(cst.children.myOtherLabel[0].name).to.equal("bamba")
-            expect(cst.children.myOtherLabel[0].recoveredNode).to.be.true
+
+            const bamba = cst.children.myOtherLabel[0] as CstNode
+            expect(bamba.name).to.equal("bamba")
+            expect(bamba.recoveredNode).to.be.true
         })
 
         it("Can output a CST for a Terminal - alternations", () => {
-            class CstTerminalAlternationParser extends Parser {
+            class CstTerminalAlternationParser extends CstParser {
                 constructor(input: IToken[] = []) {
                     super(ALL_TOKENS, {
                         outputCst: true,
@@ -174,7 +176,7 @@ function defineTestSuit(recoveryMode) {
         })
 
         it("Can output a CST for a Terminal - alternations - single", () => {
-            class CstTerminalAlternationSingleAltParser extends Parser {
+            class CstTerminalAlternationSingleAltParser extends CstParser {
                 constructor(input: IToken[] = []) {
                     super(ALL_TOKENS, {
                         outputCst: true,
@@ -206,7 +208,7 @@ function defineTestSuit(recoveryMode) {
         })
 
         it("Can output a CST for a Terminal with multiple occurrences", () => {
-            class CstMultiTerminalParser extends Parser {
+            class CstMultiTerminalParser extends CstParser {
                 constructor(input: IToken[] = []) {
                     super(ALL_TOKENS, {
                         outputCst: true,
@@ -239,7 +241,7 @@ function defineTestSuit(recoveryMode) {
         })
 
         it("Can output a CST for a Terminal with multiple occurrences - iteration", () => {
-            class CstMultiTerminalWithManyParser extends Parser {
+            class CstMultiTerminalWithManyParser extends CstParser {
                 constructor(input: IToken[] = []) {
                     super(ALL_TOKENS, {
                         outputCst: true,
@@ -281,19 +283,18 @@ function defineTestSuit(recoveryMode) {
             expect(tokenStructuredMatcher(cst.children.A[2], A)).to.be.true
             expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be.true
             expect(cst.children.bamba).to.have.length(3)
-            expect(
-                tokenStructuredMatcher(cst.children.bamba[0].children.C[0], C)
-            ).to.be.true
-            expect(
-                tokenStructuredMatcher(cst.children.bamba[1].children.C[0], C)
-            ).to.be.true
-            expect(
-                tokenStructuredMatcher(cst.children.bamba[2].children.C[0], C)
-            ).to.be.true
+
+            const children = cst.children.bamba as CstNode[]
+            expect(tokenStructuredMatcher(children[0].children.C[0], C)).to.be
+                .true
+            expect(tokenStructuredMatcher(children[1].children.C[0], C)).to.be
+                .true
+            expect(tokenStructuredMatcher(children[2].children.C[0], C)).to.be
+                .true
         })
 
         context("Can output a CST for an optional terminal", () => {
-            class CstOptionalTerminalParser extends Parser {
+            class CstOptionalTerminalParser extends CstParser {
                 constructor(input: IToken[] = []) {
                     super(ALL_TOKENS, {
                         outputCst: true,
@@ -327,13 +328,11 @@ function defineTestSuit(recoveryMode) {
                 expect(cst.name).to.equal("ruleWithOptional")
                 expect(cst.children).to.have.keys("A", "B", "bamba")
                 expect(tokenStructuredMatcher(cst.children.A[0], A)).to.be.true
-                expect(cst.children.bamba[0].name).to.equal("bamba")
-                expect(
-                    tokenStructuredMatcher(
-                        cst.children.bamba[0].children.C[0],
-                        C
-                    )
-                ).to.be.true
+
+                const bamba = cst.children.bamba[0] as CstNode
+                expect(bamba.name).to.equal("bamba")
+                expect(tokenStructuredMatcher(bamba.children.C[0], C)).to.be
+                    .true
                 expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be.true
             })
 
@@ -350,7 +349,7 @@ function defineTestSuit(recoveryMode) {
         })
 
         it("Can output a CST for a Terminal with multiple occurrences - iteration mandatory", () => {
-            class CstMultiTerminalWithAtLeastOneParser extends Parser {
+            class CstMultiTerminalWithAtLeastOneParser extends CstParser {
                 constructor(input: IToken[] = []) {
                     super(ALL_TOKENS, {
                         outputCst: true,
@@ -386,7 +385,7 @@ function defineTestSuit(recoveryMode) {
         })
 
         it("Can output a CST for a Terminal with multiple occurrences - iteration SEP", () => {
-            class CstMultiTerminalWithManySepParser extends Parser {
+            class CstMultiTerminalWithManySepParser extends CstParser {
                 constructor(input: IToken[] = []) {
                     super(ALL_TOKENS, {
                         outputCst: true,
@@ -427,7 +426,7 @@ function defineTestSuit(recoveryMode) {
         })
 
         it("Can output a CST for a Terminal with multiple occurrences - iteration SEP mandatory", () => {
-            class CstMultiTerminalWithAtLeastOneSepParser extends Parser {
+            class CstMultiTerminalWithAtLeastOneSepParser extends CstParser {
                 constructor(input: IToken[] = []) {
                     super(ALL_TOKENS, {
                         outputCst: true,
@@ -468,7 +467,7 @@ function defineTestSuit(recoveryMode) {
         })
 
         it("Can output a CST with node full location information", () => {
-            class CstTerminalParser extends Parser {
+            class CstTerminalParser extends CstParser {
                 constructor(input: IToken[] = []) {
                     super(ALL_TOKENS, {
                         nodeLocationTracking: "full",
@@ -515,13 +514,11 @@ function defineTestSuit(recoveryMode) {
             )
             expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be.true
             expect(tokenStructuredMatcher(cst.children.C[0], C)).to.be.true
-            expect(cst.children.first[0].name).to.equal("first")
-            expect(cst.children.second[0].name).to.equal("second")
-            expect(
-                tokenStructuredMatcher(cst.children.first[0].children.A[0], A)
-            ).to.be.true
 
-            expect(cst.children.first[0].location).to.deep.equal({
+            const first = cst.children.first[0] as CstNode
+            expect(first.name).to.equal("first")
+            expect(tokenStructuredMatcher(first.children.A[0], A)).to.be.true
+            expect(first.location).to.deep.equal({
                 startOffset: 1,
                 startLine: 1,
                 startColumn: 1,
@@ -530,22 +527,24 @@ function defineTestSuit(recoveryMode) {
                 endColumn: 2
             })
 
-            expect(cst.children.empty[0].location).to.deep.equal({
-                startOffset: NaN,
-                startLine: NaN,
-                startColumn: NaN,
-                endOffset: NaN,
-                endLine: NaN,
-                endColumn: NaN
-            })
-
-            expect(cst.children.second[0].location).to.deep.equal({
+            const second = cst.children.second[0] as CstNode
+            expect(second.name).to.equal("second")
+            expect(second.location).to.deep.equal({
                 startOffset: 17,
                 startLine: 5,
                 startColumn: 2,
                 endOffset: 18,
                 endLine: 5,
                 endColumn: 4
+            })
+
+            expect((cst.children.empty[0] as CstNode).location).to.deep.equal({
+                startOffset: NaN,
+                startLine: NaN,
+                startColumn: NaN,
+                endOffset: NaN,
+                endLine: NaN,
+                endColumn: NaN
             })
 
             expect(cst.location).to.deep.equal({
@@ -559,7 +558,7 @@ function defineTestSuit(recoveryMode) {
         })
 
         it("Can output a CST with node onlyOffset location information", () => {
-            class CstTerminalParser extends Parser {
+            class CstTerminalParser extends CstParser {
                 constructor(input: IToken[] = []) {
                     super(ALL_TOKENS, {
                         nodeLocationTracking: "onlyOffset",
@@ -606,23 +605,23 @@ function defineTestSuit(recoveryMode) {
             )
             expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be.true
             expect(tokenStructuredMatcher(cst.children.C[0], C)).to.be.true
-            expect(cst.children.first[0].name).to.equal("first")
-            expect(cst.children.second[0].name).to.equal("second")
-            expect(
-                tokenStructuredMatcher(cst.children.first[0].children.A[0], A)
-            ).to.be.true
 
-            expect(cst.children.first[0].location).to.deep.equal({
+            const first = cst.children.first[0] as CstNode
+            expect(first.name).to.equal("first")
+            expect(tokenStructuredMatcher(first.children.A[0], A)).to.be.true
+            expect(first.location).to.deep.equal({
                 startOffset: 1,
                 endOffset: 2
             })
 
-            expect(cst.children.second[0].location).to.deep.equal({
+            const second = cst.children.second[0] as CstNode
+            expect(second.name).to.equal("second")
+            expect(second.location).to.deep.equal({
                 startOffset: 17,
                 endOffset: 18
             })
 
-            expect(cst.children.empty[0].location).to.deep.equal({
+            expect((cst.children.empty[0] as CstNode).location).to.deep.equal({
                 startOffset: NaN,
                 endOffset: NaN
             })
@@ -634,7 +633,7 @@ function defineTestSuit(recoveryMode) {
         })
 
         it("Can output a CST with no location information", () => {
-            class CstTerminalParser extends Parser {
+            class CstTerminalParser extends CstParser {
                 constructor(input: IToken[] = []) {
                     super(ALL_TOKENS, {
                         nodeLocationTracking: "none",
@@ -672,22 +671,23 @@ function defineTestSuit(recoveryMode) {
             expect(cst.children).to.have.keys("B", "C", "first", "second")
             expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be.true
             expect(tokenStructuredMatcher(cst.children.C[0], C)).to.be.true
-            expect(cst.children.first[0].name).to.equal("first")
-            expect(cst.children.second[0].name).to.equal("second")
-            expect(
-                tokenStructuredMatcher(cst.children.first[0].children.A[0], A)
-            ).to.be.true
 
-            expect(cst.children.first[0].location).to.be.undefined
+            const first = cst.children.first[0] as CstNode
+            expect(first.name).to.equal("first")
+            expect(tokenStructuredMatcher(first.children.A[0], A)).to.be.true
 
-            expect(cst.children.second[0].location).to.be.undefined
+            expect(first.location).to.be.undefined
+
+            const second = cst.children.second[0] as CstNode
+            expect(second.name).to.equal("second")
+            expect(second.location).to.be.undefined
 
             expect(cst.location).to.be.undefined
         })
 
         context("nested rules", () => {
             context("Can output cst when using OPTION", () => {
-                class CstOptionalNestedTerminalParser extends Parser {
+                class CstOptionalNestedTerminalParser extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             outputCst: true,
@@ -726,22 +726,19 @@ function defineTestSuit(recoveryMode) {
                     let cst = parser.ruleWithOptional()
                     expect(cst.name).to.equal("ruleWithOptional")
                     expect(cst.children).to.have.keys("$nestedOption", "B")
-                    let $nestedOptionCst = cst.children.$nestedOption[0]
+                    let $nestedOptionCst = cst.children
+                        .$nestedOption[0] as CstNode
                     expect(
                         tokenStructuredMatcher(
                             $nestedOptionCst.children.A[0],
                             A
                         )
                     ).to.be.true
-                    expect($nestedOptionCst.children.bamba[0].name).to.equal(
-                        "bamba"
-                    )
-                    expect(
-                        tokenStructuredMatcher(
-                            $nestedOptionCst.children.bamba[0].children.C[0],
-                            C
-                        )
-                    ).to.be.true
+
+                    const bamba = $nestedOptionCst.children.bamba[0] as CstNode
+                    expect(bamba.name).to.equal("bamba")
+                    expect(tokenStructuredMatcher(bamba.children.C[0], C)).to.be
+                        .true
                     expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be
                         .true
                 })
@@ -753,7 +750,8 @@ function defineTestSuit(recoveryMode) {
                     expect(cst.name).to.equal("ruleWithOptional")
                     // nested rule is equivalent to an optionally empty rule call
                     expect(cst.children).to.have.keys("B", "$nestedOption")
-                    let $nestedOptionCst = cst.children.$nestedOption[0]
+                    let $nestedOptionCst = cst.children
+                        .$nestedOption[0] as CstNode
                     expect($nestedOptionCst.children.A).to.be.undefined
                     expect($nestedOptionCst.children.bamba).to.be.undefined
                     expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be
@@ -762,7 +760,7 @@ function defineTestSuit(recoveryMode) {
             })
 
             it("Can output a CST when using OR with nested named Alternatives", () => {
-                class CstAlternationNestedAltParser extends Parser {
+                class CstAlternationNestedAltParser extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             outputCst: true,
@@ -799,7 +797,7 @@ function defineTestSuit(recoveryMode) {
                 let cst = parser.testRule()
                 expect(cst.name).to.equal("testRule")
                 expect(cst.children).to.have.keys("$first_alternative")
-                let firstAltCst = cst.children.$first_alternative[0]
+                let firstAltCst = cst.children.$first_alternative[0] as CstNode
                 expect(tokenStructuredMatcher(firstAltCst.children.A[0], A)).to
                     .be.true
                 expect(cst.children.bamba).to.be.undefined
@@ -807,7 +805,7 @@ function defineTestSuit(recoveryMode) {
             })
 
             it("Can output a CST when using OR", () => {
-                class CstAlternationNestedParser extends Parser {
+                class CstAlternationNestedParser extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             outputCst: true,
@@ -846,7 +844,7 @@ function defineTestSuit(recoveryMode) {
                 let cst = parser.testRule()
                 expect(cst.name).to.equal("testRule")
                 expect(cst.children).to.have.keys("$nestedOr")
-                let orCst = cst.children.$nestedOr[0]
+                let orCst = cst.children.$nestedOr[0] as CstNode
                 expect(orCst.children).to.have.keys("A")
                 expect(tokenStructuredMatcher(orCst.children.A[0], A)).to.be
                     .true
@@ -855,7 +853,7 @@ function defineTestSuit(recoveryMode) {
             })
 
             it("Can output a CST when using OR - single Alt", () => {
-                class CstAlternationNestedAltSingleParser extends Parser {
+                class CstAlternationNestedAltSingleParser extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             outputCst: true,
@@ -887,15 +885,17 @@ function defineTestSuit(recoveryMode) {
                 let cst = parser.testRule()
                 expect(cst.name).to.equal("testRule")
                 expect(cst.children).to.have.keys("$nestedAlt")
-                let altCst = cst.children.$nestedAlt[0]
+                let altCst = cst.children.$nestedAlt[0] as CstNode
                 expect(altCst.children).to.have.keys("B", "bamba")
                 expect(tokenStructuredMatcher(altCst.children.B[0], B)).to.be
                     .true
-                expect(altCst.children.bamba[0].children).to.have.keys("C")
+                expect(
+                    (altCst.children.bamba[0] as CstNode).children
+                ).to.have.keys("C")
             })
 
             it("Can output a CST using Repetitions", () => {
-                class CstMultiTerminalWithManyNestedParser extends Parser {
+                class CstMultiTerminalWithManyNestedParser extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             outputCst: true,
@@ -935,7 +935,7 @@ function defineTestSuit(recoveryMode) {
                 expect(cst.name).to.equal("testRule")
                 expect(cst.children).to.have.keys("B", "$nestedMany")
                 expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be.true
-                let nestedManyCst = cst.children.$nestedMany[0]
+                let nestedManyCst = cst.children.$nestedMany[0] as CstNode
                 expect(nestedManyCst.children).to.have.keys("A", "bamba")
 
                 expect(nestedManyCst.children.A).to.have.length(3)
@@ -946,29 +946,18 @@ function defineTestSuit(recoveryMode) {
                 expect(tokenStructuredMatcher(nestedManyCst.children.A[2], A))
                     .to.be.true
 
-                expect(nestedManyCst.children.bamba).to.have.length(3)
-                expect(
-                    tokenStructuredMatcher(
-                        nestedManyCst.children.bamba[0].children.C[0],
-                        C
-                    )
-                ).to.be.true
-                expect(
-                    tokenStructuredMatcher(
-                        nestedManyCst.children.bamba[1].children.C[0],
-                        C
-                    )
-                ).to.be.true
-                expect(
-                    tokenStructuredMatcher(
-                        nestedManyCst.children.bamba[2].children.C[0],
-                        C
-                    )
-                ).to.be.true
+                const bamba = nestedManyCst.children.bamba as CstNode[]
+                expect(bamba).to.have.length(3)
+                expect(tokenStructuredMatcher(bamba[0].children.C[0], C)).to.be
+                    .true
+                expect(tokenStructuredMatcher(bamba[1].children.C[0], C)).to.be
+                    .true
+                expect(tokenStructuredMatcher(bamba[2].children.C[0], C)).to.be
+                    .true
             })
 
             it("Can output a CST using mandatory Repetitions", () => {
-                class CstAtLeastOneNestedParser extends Parser {
+                class CstAtLeastOneNestedParser extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             outputCst: true,
@@ -1000,7 +989,7 @@ function defineTestSuit(recoveryMode) {
                 expect(cst.name).to.equal("testRule")
                 expect(cst.children).to.have.keys("$oops", "B")
                 expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be.true
-                let oopsCst = cst.children.$oops[0]
+                let oopsCst = cst.children.$oops[0] as CstNode
                 expect(oopsCst.children).to.have.keys("A")
                 expect(oopsCst.children.A).to.have.length(3)
                 expect(tokenStructuredMatcher(oopsCst.children.A[0], A)).to.be
@@ -1012,7 +1001,7 @@ function defineTestSuit(recoveryMode) {
             })
 
             it("Can output a CST using Repetitions with separator", () => {
-                class CstNestedRuleWithManySepParser extends Parser {
+                class CstNestedRuleWithManySepParser extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             outputCst: true,
@@ -1045,7 +1034,7 @@ function defineTestSuit(recoveryMode) {
                 expect(cst.name).to.equal("testRule")
                 expect(cst.children).to.have.keys("$pizza", "B")
                 expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be.true
-                let pizzaCst = cst.children.$pizza[0]
+                let pizzaCst = cst.children.$pizza[0] as CstNode
                 expect(pizzaCst.children.A).to.have.length(2)
                 expect(tokenStructuredMatcher(pizzaCst.children.A[0], A)).to.be
                     .true
@@ -1057,7 +1046,7 @@ function defineTestSuit(recoveryMode) {
             })
 
             it("Can output a CST using Repetitions with separator - mandatory", () => {
-                class CstAtLeastOneSepNestedParser extends Parser {
+                class CstAtLeastOneSepNestedParser extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             outputCst: true,
@@ -1091,7 +1080,7 @@ function defineTestSuit(recoveryMode) {
                 expect(cst.children).to.have.keys("$nestedName", "B")
                 expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be.true
 
-                let nestedCst = cst.children.$nestedName[0]
+                let nestedCst = cst.children.$nestedName[0] as CstNode
                 expect(nestedCst.children.A).to.have.length(2)
                 expect(tokenStructuredMatcher(nestedCst.children.A[0], A)).to.be
                     .true
@@ -1103,7 +1092,7 @@ function defineTestSuit(recoveryMode) {
             })
 
             it("Can output a CST with nested rules with node full location information", () => {
-                class CstTerminalParser extends Parser {
+                class CstTerminalParser extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             nodeLocationTracking: "full",
@@ -1134,7 +1123,7 @@ function defineTestSuit(recoveryMode) {
                 expect(cst.children).to.have.keys("$nestedOption", "B")
                 expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be.true
 
-                let nestedCst = cst.children.$nestedOption[0]
+                let nestedCst = cst.children.$nestedOption[0] as CstNode
                 expect(nestedCst.children.A).to.have.length(1)
                 expect(tokenStructuredMatcher(nestedCst.children.A[0], A)).to.be
                     .true
@@ -1159,7 +1148,7 @@ function defineTestSuit(recoveryMode) {
             })
 
             it("Can output a CST with nested rules with node onlyOffset location information", () => {
-                class CstTerminalParser extends Parser {
+                class CstTerminalParser extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             nodeLocationTracking: "onlyOffset",
@@ -1190,7 +1179,7 @@ function defineTestSuit(recoveryMode) {
                 expect(cst.children).to.have.keys("$nestedOption", "B")
                 expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be.true
 
-                let nestedCst = cst.children.$nestedOption[0]
+                let nestedCst = cst.children.$nestedOption[0] as CstNode
                 expect(nestedCst.children.A).to.have.length(1)
                 expect(tokenStructuredMatcher(nestedCst.children.A[0], A)).to.be
                     .true
@@ -1207,7 +1196,7 @@ function defineTestSuit(recoveryMode) {
             })
 
             it("Can output a CST with nested rules with no node location information", () => {
-                class CstTerminalParser extends Parser {
+                class CstTerminalParser extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             nodeLocationTracking: "none",
@@ -1238,7 +1227,7 @@ function defineTestSuit(recoveryMode) {
                 expect(cst.children).to.have.keys("$nestedOption", "B")
                 expect(tokenStructuredMatcher(cst.children.B[0], B)).to.be.true
 
-                let nestedCst = cst.children.$nestedOption[0]
+                let nestedCst = cst.children.$nestedOption[0] as CstNode
                 expect(nestedCst.children.A).to.have.length(1)
                 expect(tokenStructuredMatcher(nestedCst.children.A[0], A)).to.be
                     .true
@@ -1249,7 +1238,7 @@ function defineTestSuit(recoveryMode) {
             })
 
             it("Can output a CST with nested rules with no node location information", () => {
-                class CstTerminalParser extends Parser {
+                class CstTerminalParser extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             nodeLocationTracking: undefined,
@@ -1266,7 +1255,7 @@ function defineTestSuit(recoveryMode) {
 
         context("Error Recovery", () => {
             it("re-sync recovery", () => {
-                class CstRecoveryParserReSync extends Parser {
+                class CstRecoveryParserReSync extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             outputCst: true,
@@ -1332,7 +1321,7 @@ function defineTestSuit(recoveryMode) {
 
                 let firstCollection = cst.children.first
                 expect(firstCollection).to.have.lengthOf(1)
-                let first = firstCollection[0]
+                let first = firstCollection[0] as CstNode
                 expect(first.recoveredNode).to.be.true
                 expect(first.children).to.have.keys("A")
                 expect(tokenStructuredMatcher(first.children.A[0], A)).to.be
@@ -1341,7 +1330,7 @@ function defineTestSuit(recoveryMode) {
 
                 let secondCollection = cst.children.second
                 expect(secondCollection).to.have.lengthOf(1)
-                let second = secondCollection[0]
+                let second = secondCollection[0] as CstNode
                 expect(second.recoveredNode).to.be.undefined
                 expect(second.children).to.have.keys("C", "D")
                 expect(tokenStructuredMatcher(second.children.C[0], C)).to.be
@@ -1351,7 +1340,7 @@ function defineTestSuit(recoveryMode) {
             })
 
             it("re-sync recovery nested", () => {
-                class CstRecoveryParserReSyncNested extends Parser {
+                class CstRecoveryParserReSyncNested extends CstParser {
                     constructor(input: IToken[] = []) {
                         super(ALL_TOKENS, {
                             outputCst: true,
@@ -1420,10 +1409,10 @@ function defineTestSuit(recoveryMode) {
 
                 let firstRootCollection = cst.children.first_root
                 expect(firstRootCollection).to.have.lengthOf(1)
-                let firstRoot = firstRootCollection[0]
+                let firstRoot = firstRootCollection[0] as CstNode
                 expect(firstRoot.children).to.have.keys("first")
 
-                let first = firstRoot.children.first[0]
+                let first = firstRoot.children.first[0] as CstNode
                 expect(first.recoveredNode).to.be.true
                 expect(first.children).to.have.keys("A")
                 expect(tokenStructuredMatcher(first.children.A[0], A)).to.be
@@ -1432,7 +1421,7 @@ function defineTestSuit(recoveryMode) {
 
                 let secondCollection = cst.children.second
                 expect(secondCollection).to.have.lengthOf(1)
-                let second = secondCollection[0]
+                let second = secondCollection[0] as CstNode
                 expect(second.recoveredNode).to.be.undefined
                 expect(second.children).to.have.keys("C", "D")
                 expect(tokenStructuredMatcher(second.children.C[0], C)).to.be

--- a/packages/chevrotain/test/parse/cst_visitor_spec.ts
+++ b/packages/chevrotain/test/parse/cst_visitor_spec.ts
@@ -1,5 +1,5 @@
 import { createToken } from "../../src/scan/tokens_public"
-import { Parser } from "../../src/parse/parser/traits/parser_traits"
+import { CstParser } from "../../src/parse/parser/traits/parser_traits"
 import { createRegularToken } from "../utils/matchers"
 import { keys } from "../../src/utils/utils"
 import { IToken } from "../../api"
@@ -11,7 +11,7 @@ describe("The CSTVisitor", () => {
 
     const ALL_TOKENS = [A, B, C]
 
-    class CstTerminalParserReturnVisitor extends Parser {
+    class CstTerminalParserReturnVisitor extends CstParser {
         constructor(input: IToken[] = []) {
             super(ALL_TOKENS, { outputCst: true })
             this.performSelfAnalysis()

--- a/packages/chevrotain/test/parse/traits/perf_tracer_spec.ts
+++ b/packages/chevrotain/test/parse/traits/perf_tracer_spec.ts
@@ -1,5 +1,5 @@
 import { createToken } from "../../../src/scan/tokens_public"
-import { Parser } from "../../../src/parse/parser/traits/parser_traits"
+import { CstParser } from "../../../src/parse/parser/traits/parser_traits"
 
 let skipOnBrowser = describe
 if (typeof window !== "undefined") {
@@ -21,7 +21,7 @@ skipOnBrowser("Chevrotain's Init Performance Tracing", () => {
 
     const PlusTok = createToken({ name: "PlusTok" })
 
-    class TraceParser extends Parser {
+    class TraceParser extends CstParser {
         constructor(traceInitVal) {
             super([PlusTok], {
                 outputCst: false,

